Libraries/LibWeb/DOM/Element.cpp

```diff
@@ -2430,6 +2430,64 @@ bool Element::check_visibility(Optional<CheckVisibilityOptions> options)
     return true;
 }
 
+// https://drafts.csswg.org/css-contain/#proximity-to-the-viewport
+void Element::determine_proximity_to_the_viewport()
+{
+    // An element that has content-visibility: auto is in one of three states when it comes to its proximity to the viewport:
+
+    // - The element is close to the viewport: In this state, the element is considered "on-screen": its paint
+    //   containment box's overflow clip edge intersects with the viewport, or a user-agent defined margin around the
+    //   viewport.
+    auto viewport_rect = document().viewport_rect();
+    // NOTE: This margin is meant to allow the user agent to begin preparing for an element to be in the
+    // viewport soon. A margin of 50% is suggested as a reasonable default.
+    viewport_rect.inflate(viewport_rect.width(), viewport_rect.height());
+    // FIXME: We don't have paint containment or the overflop clip edge yet, so this is just using the absolute rect for now.
+    if (paintable_box()->absolute_rect().intersects(viewport_rect))
+        m_proximity_to_the_viewport = ProximityToTheViewport::CloseToTheViewport;
+
+    // FIXME: If a filter (see [FILTER-EFFECTS-1]) with non local effects includes the element as part of its input, the user
+    //        agent should also treat the element as relevant to the user when the filter’s output can affect the rendering
+    //        within the viewport (or within the user-agent defined margin around the viewport), even if the element itself is
+    //        still off-screen.
+
+    // - The element is far away from the viewport: In this state, the element’s proximity to the viewport has been
+    //   computed and is not close to the viewport.
+    m_proximity_to_the_viewport = ProximityToTheViewport::FarAwayFromTheViewport;
+}
+
+// https://drafts.csswg.org/css-contain/#relevant-to-the-user
+bool Element::is_relevant_to_the_user()
+{
+    // An element is relevant to the user if any of the following conditions are true:
+
+    // The element is close to the viewport.
+    if (m_proximity_to_the_viewport == ProximityToTheViewport::CloseToTheViewport)
+        return true;
+
+    // Either the element or its contents are focused, as described in the focus section of the HTML spec.
+    auto* focused_element = document().focused_element();
+    if (focused_element && is_inclusive_ancestor_of(*focused_element))
+        return true;
+
+    // FIXME: Either the element or its contents are selected, where selection is described in the selection API.
+
+    // Either the element or its contents are placed in the top layer.
+    auto top_layer_check = [&](auto& element) {
+        if (element.in_top_layer())
+            return TraversalDecision::Break;
+
+        return TraversalDecision::Continue;
+    };
+    if (for_each_in_inclusive_subtree_of_type<Element>(top_layer_check) == TraversalDecision::Break)
```

Comment:
Not your issue, but this API (checking for which TraversalDecision was returned) is a little cursed.

Reply:
I agree, that is cursed.
I've changed it now to just allocate a bool before the loop and to then to check it after.